File: remove_duplicates_linked_list/removeDuplicates.go
package main

import "fmt"

type List struct {
	Value	int
	Next	*List
}

func (l *List) addMany(values []int) *List {
	current := l
	for current.Next != nil {
		current = current.Next
	}
	for _, v := range values {
		current.Next = &List{Value: v, Next: nil}
		current = current.Next
	}
	return l
}

func (l *List) getNodesArray() []int {
	var res []int

	current := l
	for current != nil {
		res = append(res, current.Value)
		current = current.Next
	}
	return res
}

func removeDuplicates(l *List) {
	current := l
	for current != nil {
		next := current.Next
		for next != nil && next.Value == current.Value {
			next = next.Next
		}
		current.Next = next
		current = current.Next
	}
}

func main() {
	test := List{Value: 1, Next: nil}
	test.addMany([]int{1, 3, 4, 4, 4, 5, 6, 6})
	fmt.Println(test.getNodesArray())
	expected := List{Value: 1, Next: nil}
	expected.addMany([]int{3, 4, 5, 6})
	fmt.Println(expected.getNodesArray())
	removeDuplicates(&test)
	fmt.Println(test.getNodesArray(), "vs", expected.getNodesArray())
}

File: remove_duplicates_linked_list/remove_duplicates.py
# Complexity O(n) - n-number of nodes in list,
# Space O(1)
class LinkedList:
    def __init__(self, value):
        self.value = value
        self.next = None

    def addMany(self, values):
        current = self
        while current.next is not None:
            current = current.next
        for value in values:
            current.next = LinkedList(value)
            current = current.next
        return self

    def getNodesInArray(self):
        nodes = []
        current = self
        while current is not None:
            nodes.append(current.value)
            current = current.next
        return nodes


def removeDuplicatesFromLinkedList(linkedList):
    curr = linkedList
    while curr:
        nextDist = curr.next
        while nextDist and nextDist.value == curr.value:
            nextDist = nextDist.next
        curr.next = nextDist
        curr = nextDist
    return linkedList


if __name__ == "__main__":
    test = LinkedList(1).addMany([1, 3, 4, 4, 4, 5, 6, 6])
    expected = LinkedList(1).addMany([3, 4, 5, 6])
    actual = removeDuplicatesFromLinkedList(test)
    print(actual.getNodesInArray(), "vs", expected.getNodesInArray())
